Automattic/woocommerce-payments
# Replace overwriting $current_section with overwriting Payments tab URL

Fixes #1646.

#### Changes proposed in this Pull Request



Okay! So first, a legend:

![Markup on 2021-05-31 at 17:34:08](https://user-images.githubusercontent.com/1759681/120217643-efad4800-c238-11eb-9d15-44d171fd965f.png)

#### Why is this happening

In #1580 we implemented a change to what the default section of the Payments tab is. It is based on the observation that when section links are rendered, the default one always has the `section` query parameter set to an empty string: `&section=` (this is the case in all tabs - see e.g. the Shipping tab's "Shipping zones" section link). This is opposed to the link on the tab itself, from which the `&section` parameter is absent.

As a result, we currently handle three possible situations:

1. `/admin.php?page=wc-settings&tab=checkout` (= link on the "Payments" tab).
    * The section parameter is absent from the URL (`! isset( $_REQUEST['section'] )`).
    * If we detect the `section` parameter is absent, we redirect to the WCPay settings page by tweaking the `$current_section` global variable under the hood.
1. `/admin.php?page=wc-settings&tab=checkout&section=` (= link on the "All payment methods" section).
    * The section parameter is present but empty (`$_REQUEST['section'] === ''`).
    * All default sections have this URL structure so we treat this as a direct request for the "All payment methods" section and do no rewriting.
1. `/admin.php?page=wc-settings&tab=checkout&section=<some section name>`
    * The section parameter has a specific value. No rewriting done because a specific section is requested.

The issue reported here demonstrates that relying on this behavior (distinguishing if `section` is empty/absent) doesn't support all links to the default WC settings page. Some links (such as the back arrow in question) have no `section` param, which is against the above assumption.

#### Proposed approach

We could try to fix the specific back arrow link mentioned in this issue but we might risk running into this issue again with another link somewhere so I'd like to suggest an alternative approach. Instead of redirecting to the WCPay section by default, let's replace the URL on the "Payments" tab button. So instead of the current:

```
/admin.php?page=wc-settings&tab=checkout
```

The link for the tab in the screenshot would now be:

```
/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments
```

This approach is safer because we don't overwrite what's regarded as the default section, potentially breaking links depending on the "All payment methods" being the default but it results in the Payments tab link being the only one with a `&section` parameter.







#### Testing instructions

1. Enable the grouped settings feature flag `_wcpay_feature_grouped_settings`.
1. Visit the WooCommerce > Settings > Payments tab.
    * You should arrive at the "WooCommerce Payments" settings page.
1. Visit the WooCommerce > Settings > Payments > All payment methods section and select a payment method.
    * After clicking the back arrow, you should arrive at the "All payment methods" settings page.

![Markup on 2021-06-01 at 13:53:31](https://user-images.githubusercontent.com/1759681/120318962-c47e3380-c2e0-11eb-8771-794a07c7f90c.png)






*

-------------------

    - Not added - feature behind feature flag.
